### ISSUE TITLE
NLL: turn off migration mode

### ISSUE THREAD

**Issue by pnkfelix** [rust-lang/rust#58781]
Transition fully to NLL (instead of the migration mode currently used).

   * one currently opts into this via `#![feature(nll)]` or `-Z borrowck=mir`
   * It gets rid of AST-region inference errors, causing some code to be accepted that is rejected by migrate mode, and changing diagnostics in many cases.
   * Potential blockers: 
     * ~NLL-sound bugs that affect NLL mode without migration
       * ~#57642 (needs test for now)~
     * Diagnostic differences
       * List: https://github.com/rust-lang/rust/issues?q=is%3Aopen+is%3Aissue+label%3ANLL-diagnostics
     * Migrate compare-mode=NLL test differences to revisions (to ease stabilization PR and meticulously check diagnostic differences)
     * ~Two-phase borrows (2PB) question: #56254~
     * ~Must have crater run showing no major crates fail in practice (#60680)~
     * #73154
       * Likely needs changes to the leak check
     * ~Investigate #96331~
       * Diagnostics-only change, should be fine

**Comment by pnkfelix** [comment]
Explicitly marking this P-medium for now, to reflect the fact that transitioning the 2015 edition to migrate mode is higher priority for the short term.

**Comment by nikomatsakis** [comment]
Created https://github.com/rust-lang/rust/issues/60680 as an issue to do a first crater run to get a rough idea.

**Comment by jethrogb** [comment]
Is this the right place for comments/questions about the migration warning?

I got this warning:
```
   = warning: this error has been downgraded to a warning for backwards compatibility with previous releases
   = warning: this represents potential undefined behavior in your code and this warning will become a hard error in the future
```

The warning is good, my code definitely has UB. I found that I can't upgrading the warning to an error with `deny(warnings)`. Why is that?

This code compiles and has UB on Rust 1.0, I thought NLL is only supposed to relax things? Are you fixing any bugs in the old borrow checker?

**Comment by pnkfelix** [comment]
@jethrogb no, NLL was not meant to solely relax things. Such relaxation is the primary selling point of NLL, but we always knew it was going to introduce breakage.

The shift to NLL is coupled with a change in the underlying borrow-checker implementation to analyze MIR (instead of analyzing the higher-level AST, which is what the old borrow checker did). In order to avoid injecting breaking changes that people could not work around in the short term, we currently have NLL in a migration mode. That was added in PR #52681.

In hindsight, it may have been nice to leverage the lint system here to let people upgrade these warnings to errors. (One reason we didn't do that is that we didn't want people to `#![allow(..)]` the warnings here. For more on that, see also issue #34596 and PR #59658.)

**Comment by Mark-Simulacrum** [comment]
@matthewjasper Could you either a) close this issue or b) update it with the current status? I *think* we've fully dropped migration mode now and even cleaned up some of the leftover AST borrowck-related bits, but there is some stuff left that is not yet fully migrated I believe.

**Comment by matthewjasper** [comment]
Migrate mode still exists, although it's not really migrating anything. I've updated the description for the current status.

**Comment by jonas-schievink** [comment]
What's the difference between this issue and https://github.com/rust-lang/rust/issues/57895?

**Comment by Aaron1011** [comment]
Are any changes needed w.r.t https://github.com/rust-lang/rust/issues/59159 before we turn off migrate mode? We currently have a lint - are there any problems with just keeping this lint when migrate mode is disabled?

**Comment by Mark-Simulacrum** [comment]
See discussion in https://github.com/rust-lang/rust/pull/76104 -- the current thinking in my mind (and I think maybe partial consensus amongst lang, though that's less clear) is that we should set the lint to allow by default or perhaps even remove it entirely.

**Comment by nikomatsakis** [comment]
The lint is orthogonal from migration mode, in any case.

**Comment by bstrie** [comment]
Presumably fully transitioning to NLL would also cause the "NLL" mode tests in the UI test suite (`*.nll.stderr`, `--compare-mode=nll`) to become the default, yes? Will the pre-NLL stderr files be completely obsolete in that case?

**Comment by matthewjasper** [comment]
Yes, this would remove `--compare-mode-nll` and make the `.nll.stderr` files replace the `.stderr` files where they exist.

**Comment by Aaron1011** [comment]
What's the current status of this issue?

**Comment by matthewjasper** [comment]
It's still blocked on #57374 and #73154. I've started working on #57374.

**Comment by Aaron1011** [comment]
@matthewjasper: Are you still working on https://github.com/rust-lang/rust/issues/57374? If not, I'd be interested in working on it

**Comment by matthewjasper** [comment]
I made some good progress (see https://github.com/matthewjasper/rust/tree/nll-hrtb-errors) but haven't had a chance to finish it. Specifically I wanted to avoid the code duplication between `bound_region_errors` and the existing code that implements those queries. I also wanted to handle "ascribe user type" queries as well.

**Comment by lqd** [comment]
@matthewjasper would it make sense to try to merge this great progress you've already completed ? or would you rather wait for the 2 tasks you mentioned to be completed before doing so ?

(If trying to land a PR of your `nll-hrtb-errors` branch soon would be interesting, I thought I'd try to help by rebasing it already, [over here](https://github.com/lqd/rust/tree/matthews-nll-hrtb-errors))

**Comment by matthewjasper** [comment]
You could try that. The branch is pretty old, but hopefully touches parts of the codebase that move pretty slowly.

**Comment by lqd** [comment]
Sorry, I meant I had already rebased your branch in the link above: it was indeed in slow-moving parts of the compiler, with only a few simple conflicts, and only one compile error. Very straightforward to update overall.

Your improvements to these error messages are already looking great. I can open a PR and help land them, if @nikomatsakis and @Aaron1011 want to have a look ?

**Comment by Aaron1011** [comment]
Now that higher-ranked subtype errors are improved for stable code (https://github.com/rust-lang/rust/issues/57374#issuecomment-932490246), is https://github.com/rust-lang/rust/issues/73154 the only blocker?

**Comment by lqd** [comment]
A few ideas:
- maybe the couple needed cleanups that we mentioned before. I had started deduplicating the normalize query in the HRTB diagnostic code and the changes are not what I expected at first (there's no failure to normalize, even though that's what the fallback error message says when the nice errors can't be emitted). It looks more like changes only impacting what this test checks with the leak check turned off.
- which reminds me, do we need to have a plan about the leak check, if we want to turn NLLs on ?
- just checking: even though it's still marked as potential blocker in some tracking issues, it _seems_ that we don't need to do anything about the 2PB situation and the `mutable_borrow_reservation_conflict` lint from #59159 to turn NLLs on ?
- we should take another look at the NLL diagnostics: both checking the existing 225 blessed NLL test expectations, and triage the [17 open NLL-Diagnostics issues](https://github.com/rust-lang/rust/issues?utf8=%E2%9C%93&q=is%3Aopen+is%3Aissue+label%3ANLL-diagnostics+) to see if they match the non-NLL diagnostics in quality and clarity.

**Comment by nikomatsakis** [comment]
Regarding the leak check: 

I believe we *do* have to reproduce some of that functionality.

**Comment by marmeladema** [comment]
Is there something actionable that a small contributor like me could help with? It would be very pleasant to finally stabilize NLL 🥳

**Comment by bstrie** [comment]
@marmeladema The diagnostic issues mentioned above could probably use triage to see if they're still relevant: https://github.com/rust-lang/rust/issues?q=is%3Aopen+is%3Aissue+label%3ANLL-diagnostics . If they are, then resolving those may be a good place to start. You can also try hopping into [#t-compiler](https://rust-lang.zulipchat.com/#narrow/stream/131828-t-compiler) or [#t-compiler/help](https://rust-lang.zulipchat.com/#narrow/stream/182449-t-compiler.2Fhelp) and asking there.

**Comment by marmeladema** [comment]
I opened a zulip threads in the wg-nll group about those: https://rust-lang.zulipchat.com/#narrow/stream/122657-t-compiler.2Fwg-nll/topic/nll.20diagnostics.20issues

TL;DR is that there are actually 4 and we believe they can all probably be closed.

**Comment by jackh726** [comment]
I've updated the op to include the latest potential blockers and added a tracking label to reflect that some impl work needs to be re. #73154.

**Comment by m-ou-se** [comment]
Do we know why turning off migration mode fixes https://github.com/rust-lang/rust/issues/95527?

Is that something that can be fixed without having to wait for turning off migration mode (to not block scoped threads on https://github.com/rust-lang/rust/issues/73154)?
